server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/InternalBinaryRange.java

```diff
@@ -57,30 +57,28 @@ public Bucket(DocValueFormat format, boolean keyed, String key, BytesRef from, B
                 long docCount, InternalAggregations aggregations) {
             this.format = format;
             this.keyed = keyed;
-            this.key = key;
+            this.key = key != null ? key : generateKey(from, to, format);
             this.from = from;
             this.to = to;
             this.docCount = docCount;
             this.aggregations = aggregations;
         }
 
-        // for serialization
-        private Bucket(StreamInput in, DocValueFormat format, boolean keyed) throws IOException {
-            this.format = format;
-            this.keyed = keyed;
-            key = in.readOptionalString();
-            if (in.readBoolean()) {
-                from = in.readBytesRef();
-            } else {
-                from = null;
-            }
-            if (in.readBoolean()) {
-                to = in.readBytesRef();
-            } else {
-                to = null;
-            }
-            docCount = in.readLong();
-            aggregations = InternalAggregations.readAggregations(in);
+        private static String generateKey(BytesRef from, BytesRef to, DocValueFormat format) {
+            StringBuilder builder = new StringBuilder()
+                .append(from == null ? "*" : format.format(from))
+                .append("-")
+                .append(to == null ? "*" : format.format(to));
+            return builder.toString();
+        }
+
+        private static Bucket createFromStream(StreamInput in, DocValueFormat format, boolean keyed) throws IOException {
+            String key = in.readOptionalString();
```

Comment:
given that the key is never null anymore, let's write it with writeString and read it with readString?

Reply:
I had opted not to change this because of the many follow-up changes required. I'm happy to do this though, just to check I have all the steps right:
- Change this to `readString` if the input stream has version >= 7.0.0 (and make the corresponding change for writing). Merge the PR.
- Backport to 6.x, changing the version check to >= 6.4.0.
- On master, also change the version check to >= 6.4.0.
- Likely do the same fix for `InternalRange` (it also uses `writeOptionalString`).

Reply:
You got the steps right. I agree this is a bit painful but I think this is worth doing otherwise we are accumulating tech debt and someone who doesn't know about the history of this class will wonder why we have an `optional` two years from now. For the record, I assumed this PR was 7.0-only based on the label, which would have made the serialization change easier to merge.

+1 to fix InternalRange as well

Reply:
Thanks, I made the change to `readString`. And makes sense about the missing PR label!